runtime: Fix stdout/stderr output from container being truncated

littlejawa opened the pull request:
Do not close the tty as part of the stdout redirection routine.
The close is already happening a couple lines below, after all routines
have finished.

Fixes #1713

Signed-off-by: Julien Ropé <jrope@redhat.com>

fidencio commented:
/test

fidencio commented:
/retest

fidencio commented:
@littlejawa, firstly, thanks for the contribution!

I'm wondering how hard was to reproduce this with a really huge file.  If we could have a reproducer for this as part of https://github.com/kata-containers/tests/tree/main/integration/kubernetes, that would be amazing!
/cc @wainersm as he can help!

littlejawa commented:
@fidencio I can certainly try.
So far, I only have a very hard-coded/specific script to help reproduce, but it probably can be improved.
I'll have a look, but any guidance is welcome :-)

lifupan commented on src/runtime/containerd-shim-v2/stream.go:
```diff
@@ -109,7 +109,6 @@ func ioCopy(exitch, stdinCloser chan struct{}, tty *ttyIO, stdinPipe io.WriteClo
 			defer bufPool.Put(p)
 			io.CopyBuffer(tty.Stdout, stdoutPipe, *p)
 			wg.Done()
-			closeOnce.Do(tty.close)
```
You can't  delete this tty close, since it will close the tty's stdin and then the following wg.Wait() can continue, otherwise, the stdin iocopy wouldn't ended and there's a dead lock occur.  The root issue here is that it shouldn't close the tty's stderr at the end of the stdout iocopy, so in addition to delete "closeOnce.Do(tty.close)",  you should also close the tty's stdin here.

littlejawa replied:
Right. In my tests, stdin was nil so the routine was not started, and I couldn't see this deadlock.
I will change that. Thanks!

littlejawa commented:
@lifupan : modified as asked.
I also removed the call to sync.Once as we're now calling tty.close() only once anyway, so no synchronization is needed - am I right?

lifupan commented:
> @lifupan : modified as asked.
> I also removed the call to sync.Once as we're now calling tty.close() only once anyway, so no synchronization is needed - am I right?

Hi @littlejawa 

You are right, thanks. It LGTM.

lifupan commented:
/test

fidencio commented:
/retest-debian
/retest-fedora

fidencio reviewed (approved):
lgtm, thanks @littlejawa!

fidencio commented:
/retest-fedora

fidencio commented:
/retest-fedora

fidencio commented:
/retest-fedora

fidencio commented:
/test-arm
